SIPS/sip-6.md

```diff
@@ -0,0 +1,184 @@
+---
+sip: 6
+title: Deterministic Snap-specific entropy
+status: Draft
+discussions-to: https://github.com/MetaMask/SIPs/discussions/69
+author: Maarten Zuidhoorn (@Mrtenz)
+created: 2022-10-27
+---
+
+## Abstract
+
+This SIP describes a way for Snaps to get some deterministic Snap-specific entropy, based on the secret recovery phrase
+of the user and the snap ID. Since Snaps do not have access to the secret recovery phrase directly, other Snaps are
+unable to get the same entropy.
+
+Snaps can optionally specify a salt in order to generate different entropy for different use cases, for example, some
+entropy to derive new private keys from, and some other entropy to encrypt some data.
+
+The entropy can be accessed from within a Snap, using the `snap_getEntropy` JSON-RPC method.
+
+## Motivation
+
+Before this SIP, Snaps did not have a way to get some kind of deterministic entropy, without storing it on the disk. If
+the user deletes the Snap, or deletes the data of MetaMask, this entropy is lost. This SIP proposes a new way to get
+deterministic entropy, which is tied to the secret recovery phrase of the user, combined with the Snap ID. In this case,
+the entropy can always be re-created, as long as the user has a copy of their secret recovery phrase.
+
+## Specification
+
+> Formal specifications are written in TypeScript.
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",
+"NOT RECOMMENDED", "MAY", and "OPTIONAL" written in uppercase in this document are to be interpreted as described in
+[RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Deriving the entropy
+
+Entropy is derived using a
+[BIP 32](https://github.com/bitcoin/bips/blob/6545b81022212a9f1c814f6ce1673e84bc02c910/bip-0032.mediawiki) derivation
+path. This derivation path MUST start with the magic value `0xd36e6170` (`1399742832'` in BIP 32 notation). The
+following eight indices are based on a hash of the Snap ID, with an optional salt. The hashing algorithm of choice is
+`keccak256`.
+
+The hash of the Snap ID is calculated as follows:
+
+```typescript
+const hash = keccak256(snapId + keccak256(salt));
+```
+
+If the salt is not provided, an empty string MUST be used instead:
+
+```typescript
+const hash = keccak256(snapId + keccak256(''));
+```
+
```

Comment:
The developer of the snap can decide whether to use a salt or not? What is the purpose of it? Why would they add a salt, or choose not to?